File: tools/setup-mirai-analytics.mjs
#!/usr/bin/env node
/**
 * 未来喫茶 — Supabase アナリティクス一括セットアップ
 *
 * 事前準備:
 *   1. https://supabase.com で無料アカウント作成
 *   2. Account → Access Tokens → Generate new token
 *   3. .env.analytics.local を作成（.env.analytics.local.example を参照）
 *
 * 実行: node tools/setup-mirai-analytics.mjs
 */
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { randomBytes } from 'crypto';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..');
const ENV_PATH = join(ROOT, '.env.analytics.local');
const SQL_PATH = join(ROOT, 'data', 'analytics-supabase.sql');
const CONFIG_PATH = join(ROOT, 'js', 'analytics-config.js');
const API = 'https://api.supabase.com/v1';

function loadEnv() {
  if (!existsSync(ENV_PATH)) {
    console.error('❌ .env.analytics.local がありません。');
    console.error('   .env.analytics.local.example をコピーして値を入れてください。');
    process.exit(1);
  }
  const env = {};
  readFileSync(ENV_PATH, 'utf8')
    .split('\n')
    .forEach((line) => {
      const t = line.trim();
      if (!t || t.startsWith('#')) return;
      const i = t.indexOf('=');
      if (i < 1) return;
      env[t.slice(0, i).trim()] = t.slice(i + 1).trim();
    });
  return env;
}

async function mgmt(token, path, options = {}) {
  const res = await fetch(API + path, {
    ...options,
    headers: {
      Authorization: 'Bearer ' + token,
      'Content-Type': 'application/json',
      'User-Agent': 'mirai-kissa-setup/1.0',
      ...(options.headers || {}),
    },
  });
  const text = await res.text();
  let data = null;
  try {
    data = text ? JSON.parse(text) : null;
  } catch {
    data = { raw: text };
  }
  if (!res.ok) {
    const msg = data?.message || data?.error || res.statusText || text;
    throw new Error(`Management API ${path} → ${res.status}: ${msg}`);
  }
  return data;
}

async function waitHealthy(token, ref, maxMin = 12) {
  const end = Date.now() + maxMin * 60 * 1000;
  while (Date.now() < end) {
    const h = await mgmt(token, `/projects/${ref}/health`);
    const db = Array.isArray(h) ? h.find((x) => x.name === 'db') : null;
    if (db?.healthy) return;
    process.stdout.write('.');
    await new Promise((r) => setTimeout(r, 8000));
  }
  throw new Error('プロジェクトの起動がタイムアウトしました。しばらく待って再実行してください。');
}

function pickKeys(apiKeys) {
  const list = Array.isArray(apiKeys) ? apiKeys : apiKeys?.api_keys || [];
  let anon = '';
  let service = '';
  for (const k of list) {
    const name = (k.name || k.type || '').toLowerCase();
    if (name === 'anon' || name === 'publishable') anon = k.api_key || k.key || '';
    if (name === 'service_role' || name === 'secret') service = k.api_key || k.key || '';
  }
  return { anon, service };
}

async function runSql(token, ref, sql) {
  const query = sql
    .split('\n')
    .filter((line) => !line.trim().startsWith('--'))
    .join('\n')
    .trim();
  await mgmt(token, `/projects/${ref}/database/query`, {
    method: 'POST',
    body: JSON.stringify({ query }),
  });
}

async function createAdminUser(projectUrl, serviceKey, email, password) {
  const res = await fetch(projectUrl.replace(/\/$/, '') + '/auth/v1/admin/users', {
    method: 'POST',
    headers: {
      apikey: serviceKey,
      Authorization: 'Bearer ' + serviceKey,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ email, password, email_confirm: true }),
  });
  const data = await res.json().catch(() => ({}));
  if (!res.ok && !/already|exists|registered/i.test(JSON.stringify(data))) {
    throw new Error('管理者ユーザー作成: ' + (data.msg || data.message || res.status));
  }
}

function writeConfig(url, anonKey) {
  const body = `/**
 * アナリティクス設定（setup-mirai-analytics.mjs で生成）
 */
window.MIRAI_ANALYTICS_CONFIG = {
  enabled: true,
  supabaseUrl: '${url}',
  supabaseAnonKey: '${anonKey}',
};
`;
  writeFileSync(CONFIG_PATH, body, 'utf8');
  console.log('✅ js/analytics-config.js を更新しました');
}

async function main() {
  const env = loadEnv();
  const token = env.SUPABASE_ACCESS_TOKEN;
  const adminEmail = env.ADMIN_EMAIL || '';
  const adminPassword = env.ADMIN_PASSWORD || '';
  if (!token) {
    console.error('❌ SUPABASE_ACCESS_TOKEN が必要です');
    process.exit(1);
  }
  const createAdmin = !!(adminEmail && adminPassword);

  let ref = env.SUPABASE_PROJECT_REF || '';
  let projectUrl = env.SUPABASE_URL || '';

  if (!ref) {
    console.log('📦 組織一覧を取得…');
    const orgs = await mgmt(token, '/organizations');
    const org = orgs?.[0];
    if (!org?.slug && !org?.id) throw new Error('組織が見つかりません。Supabase ダッシュボードで組織を作成してください。');
    const orgSlug = org.slug || org.id;
    const dbPass = env.DB_PASSWORD || randomBytes(24).toString('base64url');
    const projectName = env.PROJECT_NAME || 'mirai-kissa-analytics';
    console.log(`📦 プロジェクト「${projectName}」を作成中（数分かかります）…`);
    const created = await mgmt(token, '/projects', {
      method: 'POST',
      body: JSON.stringify({
        name: projectName,
        organization_slug: orgSlug,
        db_pass: dbPass,
        region_selection: { type: 'smartGroup', code: env.REGION || 'apac' },
      }),
    });
    ref = created.id || created.ref;
    projectUrl = 'https://' + ref + '.supabase.co';
    console.log('   ref:', ref);
    console.log('   URL:', projectUrl);
    if (!env.DB_PASSWORD) {
      console.log('   ※ DB パスワード（控えてください）:', dbPass);
    }
    process.stdout.write('⏳ DB 起動待ち');
    await waitHealthy(token, ref);
    console.log(' OK');
  } else if (!projectUrl) {
    projectUrl = 'https://' + ref + '.supabase.co';
  }

  console.log('🔑 API キー取得…');
  const keys = await mgmt(token, `/projects/${ref}/api-keys?reveal=true`);
  const { anon, service } = pickKeys(keys);
  if (!anon || !service) throw new Error('anon / service_role キーが取得できませんでした');

  console.log('🗄️ SQL 実行…');
  const sql = readFileSync(SQL_PATH, 'utf8');
  await runSql(token, ref, sql);

  if (createAdmin) {
    console.log('👤 管理者ユーザー作成…');
    await createAdminUser(projectUrl, service, adminEmail, adminPassword);
  } else {
    console.log('⏭️ ADMIN_EMAIL / ADMIN_PASSWORD 未設定 → 管理者はダッシュボードで追加してください');
  }

  writeConfig(projectUrl, anon);

  console.log('\n========================================');
  console.log('  セットアップ完了');
  console.log('========================================');
  console.log('管理者ページ: （公開URL）/#/admin');
  if (createAdmin) console.log('ログイン:', adminEmail);
  else {
    console.log('ログイン: Supabase → Authentication → Users → Add user');
    console.log('  （メール・パスワードを設定し #/admin でログイン）');
  }
  console.log('\n次: git add js/analytics-config.js && git commit && git push');
}

main().catch((err) => {
  console.error('❌', err.message || err);
  process.exit(1);
});

File: tools/setup-mirai-analytics.py
#!/usr/bin/env python3
"""未来喫茶 — Supabase アナリティクス一括セットアップ（Python版）"""
import json
import os
import re
import secrets
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
ENV_PATH = ROOT / ".env.analytics.local"
SQL_PATH = ROOT / "data" / "analytics-supabase.sql"
CONFIG_PATH = ROOT / "js" / "analytics-config.js"
API = "https://api.supabase.com/v1"


def load_env():
    if not ENV_PATH.exists():
        print("❌ .env.analytics.local がありません", file=sys.stderr)
        sys.exit(1)
    env = {}
    for line in ENV_PATH.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        env[k.strip()] = v.strip()
    return env


def mgmt(token, path, method="GET", body=None):
    url = API + path
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(
        url,
        data=data,
        method=method,
        headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
            "User-Agent": "mirai-kissa-setup/1.0",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as res:
            raw = res.read().decode()
            return json.loads(raw) if raw else None
    except urllib.error.HTTPError as e:
        raw = e.read().decode()
        try:
            err = json.loads(raw)
            msg = err.get("message") or err.get("error") or raw
        except json.JSONDecodeError:
            msg = raw or e.reason
        raise RuntimeError(f"Management API {path} → {e.code}: {msg}") from e


def wait_healthy(token, ref, max_min=12):
    end = time.time() + max_min * 60
    while time.time() < end:
        h = mgmt(token, f"/projects/{ref}/health?services=db")
        if isinstance(h, list):
            db = next((x for x in h if x.get("name") == "db"), None)
            if db and db.get("healthy"):
                return
        print(".", end="", flush=True)
        time.sleep(8)
    raise RuntimeError("プロジェクト起動タイムアウト")


def pick_keys(api_keys):
    lst = api_keys if isinstance(api_keys, list) else api_keys.get("api_keys", [])
    anon = service = ""
    for k in lst:
        name = (k.get("name") or k.get("type") or "").lower()
        key = k.get("api_key") or k.get("key") or ""
        if name in ("anon", "publishable"):
            anon = key
        if name in ("service_role", "secret"):
            service = key
    return anon, service


def run_sql(token, ref, sql_text):
    # コメント行を除き、ファイル全体を一度に実行（; 分割は CHECK 句で誤分割するため不可）
    lines = []
    for line in sql_text.splitlines():
        s = line.strip()
        if s.startswith("--"):
            continue
        lines.append(line)
    query = "\n".join(lines).strip()
    mgmt(token, f"/projects/{ref}/database/query", "POST", {"query": query})


def create_admin(project_url, service_key, email, password):
    url = project_url.rstrip("/") + "/auth/v1/admin/users"
    body = json.dumps({"email": email, "password": password, "email_confirm": True}).encode()
    req = urllib.request.Request(
        url,
        data=body,
        method="POST",
        headers={
            "apikey": service_key,
            "Authorization": f"Bearer {service_key}",
            "Content-Type": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=60) as res:
            res.read()
    except urllib.error.HTTPError as e:
        raw = e.read().decode()
        if not re.search(r"already|exists|registered", raw, re.I):
            raise RuntimeError(f"管理者ユーザー作成: {raw}") from e


def write_config(url, anon_key):
    CONFIG_PATH.write_text(
        f"""/**
 * アナリティクス設定（setup-mirai-analytics.py で生成）
 */
window.MIRAI_ANALYTICS_CONFIG = {{
  enabled: true,
  supabaseUrl: '{url}',
  supabaseAnonKey: '{anon_key}',
}};
""",
        encoding="utf-8",
    )
    print("✅ js/analytics-config.js を更新しました")


def main():
    env = load_env()
    token = env.get("SUPABASE_ACCESS_TOKEN", "")
    admin_email = env.get("ADMIN_EMAIL", "")
    admin_password = env.get("ADMIN_PASSWORD", "")
    if not token:
        print("❌ SUPABASE_ACCESS_TOKEN が必要です", file=sys.stderr)
        sys.exit(1)
    create_admin_user = bool(admin_email and admin_password)

    ref = env.get("SUPABASE_PROJECT_REF", "")
    project_url = env.get("SUPABASE_URL", "")

    if not ref:
        print("📦 組織一覧を取得…")
        orgs = mgmt(token, "/organizations")
        if not orgs:
            raise RuntimeError("組織が見つかりません")
        org_slug = orgs[0].get("slug") or orgs[0].get("id")
        db_pass = env.get("DB_PASSWORD") or secrets.token_urlsafe(24)
        name = env.get("PROJECT_NAME", "mirai-kissa-analytics")
        print(f"📦 プロジェクト「{name}」を作成中…")
        created = mgmt(
            token,
            "/projects",
            "POST",
            {
                "name": name,
                "organization_slug": org_slug,
                "db_pass": db_pass,
                "region_selection": {"type": "smartGroup", "code": env.get("REGION", "apac")},
            },
        )
        ref = created.get("id") or created.get("ref")
        project_url = f"https://{ref}.supabase.co"
        print("   ref:", ref)
        print("   URL:", project_url)
        if "DB_PASSWORD" not in env:
            print("   ※ DB パスワード（控えてください）:", db_pass)
        print("⏳ DB 起動待ち", end="", flush=True)
        wait_healthy(token, ref)
        print(" OK")
    elif not project_url:
        project_url = f"https://{ref}.supabase.co"

    print("🔑 API キー取得…")
    keys = mgmt(token, f"/projects/{ref}/api-keys?reveal=true")
    anon, service = pick_keys(keys)
    if not anon or not service:
        raise RuntimeError("anon / service_role キーが取得できませんでした")

    print("🗄️ SQL 実行…")
    run_sql(token, ref, SQL_PATH.read_text(encoding="utf-8"))

    if create_admin_user:
        print("👤 管理者ユーザー作成…")
        create_admin(project_url, service, admin_email, admin_password)
    else:
        print("⏭️ 管理者は Supabase → Authentication → Users で追加してください")

    write_config(project_url, anon)
    print("\n========================================")
    print("  セットアップ完了")
    print("========================================")
    if create_admin_user:
        print("ログイン:", admin_email)


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print("❌", e, file=sys.stderr)
        sys.exit(1)
